owon-parse.c

```diff
@@ -33,50 +33,89 @@ int main(int argc, char **argv) {
 
   struct stat stbuf;
 
+  int fileSize = 0;
+
   char *buffer;
+  char *stdinBuf;
+
   
   if (argc<2) {
     printf("Give me the food !\n");
     return 1;
   }
 
-  fd=open(argv[1],O_RDONLY);
-  if (fd==-1) {
-    printf("Error: can't open file %s\n",argv[1]);
-    return(128);
-  }
-  fp=fdopen(fd,"rb");
-  if (fp==NULL) {
-    printf("Error: can't open file %s\n",argv[1]);
-    return(128);
-  }
+  if (!strcmp(argv[1], "-")) {
+    // read from stdin
+    fp = stdin;
+    fd = 0;
 
-  if (fstat(fd, &stbuf) == -1) {
-    printf("Error: %s may not be a regular file\n",argv[1]);
-    return(127);
-  }
+    int fs;
+    int c;
 
-  buffer = calloc(stbuf.st_size,sizeof(char));
-  if (buffer==NULL) {
+    stdinBuf = calloc(1000, sizeof(char));
 
-    printf("Can't allocate %d bytes of memory.\n",stbuf.st_size);
-    return(126);
-  }
-  
+    while (EOF != (c = fgetc(fp))) {
+      if (fs % 1000 == 0) {
+        if ((stdinBuf = realloc(stdinBuf, (fs + 1000) * sizeof(char))) == NULL) {
+          printf("Error: realloc returned NULL");
+          return(42);
+        }
+      }
+      
+      stdinBuf[fs] = c;
+      fs++;
+    }
 
-  if (fread((void *)buffer,sizeof(char),stbuf.st_size,fp) != stbuf.st_size) {
-    printf("Error: can't read file %s\n",argv[1]);
-    return(125);
-  }
+    fileSize = fs;
+
+    buffer = calloc(fileSize,sizeof(char));
+    if (buffer==NULL) {
+      printf("Can't allocate %d bytes of memory.\n",fileSize);
+      return(126);
+    }
+
+    memcpy(buffer, stdinBuf, fileSize);
```

Comment:
Why copy the buffer since you already have it in stdinBuf. Are you trying to reduce the memory usage during the execution of the script because it has up to 999 chars extra?

Reply:
I think I tried using the Buffer directly first, but I could not get it working. Memory usage is not important for my use case. To be honest, I can not really remember why I did this :-S I was quite happy that the code was working and does not look to strange to me ;-) - as mentioned I am not very used to C beside some Arduino scripts, so this might result from a lack of knowledge how to do it better ;-)